src/java.desktop/share/classes/javax/swing/table/TableCellEditor.java

```diff
@@ -31,8 +31,7 @@
 
 /**
  * This interface defines the method any object that would like to be
- * an editor of values for components such as <code>JListBox</code>,
- * <code>JComboBox</code>, <code>JTree</code>, or <code>JTable</code>
+ * an editor of values for the component <code>JTable</code> that
  * needs to implement.
```

Comment:
```suggestion
 * an editor of values for the {@code JTable} component needs to implement.
```
_‘that’_ after `JTable` shouldn't be there.

However, I would suggest rephrasing the Javadoc to make easier to understand.

Reply:
I'm not sure this fix is going in the right direction. Isn't the doc saying that if you want to use one of these components as the cell of a JTable it should implement this interface ?
And the problem with JListBox is that it is actually just called JList ?

Reply:
As I read it, the interface `TableCellEditor` adds the `getTableCellEditorComponent` method to the generic `CellEditor` interface.

There's also `TreeCellEditor` interface which adds `getTreeCellEditorComponent` specifically for `JTree`. This one has a cleaner [Javadoc](https://docs.oracle.com/en/java/javase/17/docs/api/java.desktop/javax/swing/tree/TreeCellEditor.html):

> Adds to CellEditor the extensions necessary to configure an editor in a tree.

The `TableCellEditor` interface should say something similar.

Reply:
Currently, the Javadoc says, “…that would like to be an editor of values _for components_ such as JListBox, JComboBox, JTree, or JTable…” Yet the interface adds a method which is relevant for `JTable` only. `JList` has no cell editor. `JComboBox` uses its own `ComboBoxEditor` interface, and `JTree` uses `TreeCellEditor`.

So, I think we're in the right direction.

Reply:
Yeah, update the PR.

Reply:
> Yeah, update the PR.

Thanks. It looks better now.

Reply:
@prrace Shall we update `TableCellEditor` Javadoc to _“Adds to CellEditor the extensions necessary to configure an editor in a table.”_ to make it similar to `TreeCellEditor`?

Reply:
Isn't this is another way around? The doc says that the object implementing this interface can be used "as an editor of values for components such as XXX", and it does not say that the XXX class should support the CellTableEditor? 
For example, the DefaultCellEditor class can be used as an editor for JComboBox/JCheckBox/JTextField.

Reply:
Hm… It's not the other way around. `DefaultCellEditor` implements `TableCellEditor`; the interface defines the `getTableCellEditorComponent` method that returns the _real component_ which will be used as the value editor.

So yeah, you're absolutely right: the current statement in the doc seems wrong.

I'm for changing the Javadoc for `TableCellEditor` to be similar to `TreeCellEditor`: _“Adds to CellEditor the extensions necessary to configure an editor in a table.”_

The object that implements the interface is not necessarily the editor for values. `DefaultCellEditor`  is the perfect example: it's not an editor itself, it's not even a component, all it does is configures a delegate, a `JComponent`, which serves as the editor.

Reply:
> The object that implements the interface is not necessarily the editor for values. DefaultCellEditor is the perfect example: it's not an editor itself, it's not even a component, all it does is configures a delegate, a JComponent, which serves as the editor.

My thoughts about this. The object which implements the interface is always the "editor", even if it is not a component. I guess you are mixing it with the other classes like "JEditorPane/JTextField/etc". 

The current spec:

>This interface defines the method any object that would like to be
an editor of values for components such as <code>JListBox</code>,
<code>JComboBox</code>, <code>JTree</code>, or <code>JTable</code>
needs to implement.

This means that the application may have some values/data stored in the JComboBox/JTree/JTable and the application may create a cell editor by implementing the TableCellEditor. That editor then could be used as a cell editor in the "main" JTable.

So the next statement in the description of this PR is not completely right:
> "CellTableEditor Interface doesn't support JComboBox and JTree components"

The JComboBox and other classes do not have to implement/support that interface.

Reply:
Yeah, if u refer this example page [https://bip.weizmann.ac.il/course/prog2/tutorial/uiswing/components/example-1dot4/ColorEditor.java], the interface is used for a cell in a table which has a button in turn invokes "JColorChooser" class. Which means the interface is used to store a particular value while the cell in is editor mode......?

Reply:
After going on through the example, what I felt is "Adds to CellEditor the extensions necessary to configure an editor in a tree" is little generic description than the updated PR Definition.

Reply:
> > The object that implements the interface is not necessarily the editor for values. DefaultCellEditor is the perfect example: it's not an editor itself, it's not even a component, all it does is configures a delegate, a JComponent, which serves as the editor.
> 
> My thoughts about this. The object which implements the interface is always the "editor", even if it is not a component. I guess you are mixing it with the other classes like "JEditorPane/JTextField/etc".

No, I don't mix them. It's called CellEditor, however, it's actually CellEditor-provider: the implementation of the interface *provides a component* which is used as the editor for cell values. Or do I miss anything?

What I mean is that this editor object defines the implementation of how an instance of `JTable` acquires an editor component. In that sense it's an editor.

Probably, this part of the discussion is off-topic as we're discussing the terms.

Let's discuss the spec.

> The current spec:
> 
> > This interface defines the method any object that would like to be
> > an editor of values for components such as `JListBox`,
> > `JComboBox`, `JTree`, or `JTable`
> > needs to implement.
> 
> This means that the application may have some values/data stored in the JComboBox/JTree/JTable and the application may create a cell editor by implementing the TableCellEditor. That editor then could be used as a cell editor in the "main" JTable.

Why does the interface reference this? The application can store and display data in any component.

Moreover the statement in the Javadoc is wrong: This interface defines the method that can create an editor for `JTable` only. The returned component is of type `Component`, therefore it can be any component.

> So the next statement in the description of this PR is not completely right:
> 
> > "CellTableEditor Interface doesn't support JComboBox and JTree components"
> 
> The JComboBox and other classes do not have to implement/support that interface.

No, they don't. At the same time `TableCellEditor` interface can't be used to return a cell editor for `JComboBox` and for `JTree` components. These components use `ComboBoxEditor` and `TreeCellEditor` correspondingly.

Reply:
> After going on through the example, what I felt is "Adds to CellEditor the extensions necessary to configure an editor in a tree" is little generic description than the updated PR Definition.

The interface is generic, so the generic description fits well.

I still find the proposed version quite hard to parse. In my first comment, I suggested rephrasing the Javadoc to make it clearer and easier to understand.

Reply:
> No, they don't. At the same time `TableCellEditor` interface can't be used to return a cell editor for `JComboBox` and for `JTree` components. These components use `ComboBoxEditor` and `TreeCellEditor` correspondingly.

Right the TableCellEditor should not be used as a JComboBox/JTree editor, it is only for JTable. But the objects implemented TableCellEditor interface can be used to modify the values stored as the JComboBox/JTree when the user edits the cell of the table. So the user selects the cell -> the TableCellEditor is activated for editing that cell -> if the data for the cell may be stored as a list then the JComboBox could be shown -> the user may change/edit the current value of the cell using JComboBox and its values? And the list of components in the spec is just an example of what components can be used by the TableCellEditor.

At least this is how I read it, might be wrong.

Reply:
I agree it probably referred to components which can be used as an editor. `DefaultEditorCell` supports three components: `JCheckBox`, `JComboBox`, and `JTextField`.

> But the objects implemented TableCellEditor interface can be used to modify the values stored as the JComboBox/JTree when the user edits the cell of the table.

And this just adds to the confusion rather than clarifies the purpose of the interface.

Reply:
In fact, the `CellEditor` interface describes how the interface can be used for `JTable` and `JTree` and gives `JTextField`, `JCheckBox`, `JComboBox` as the examples of specific editors.

Reply:
Hi, I have updated the PR as per latest discussions and Input, any other suggestions or Inputs......?